### PR TITLE
Always run unlock script completely even when cancelled or failed

### LINKED ISSUE #6: Handling workflow/job cancelling
Is it possible to run the post run step even if the action was cancelled? I have had to delete the generated branches and restart all stuck jobs because of cancelling a running job.

### PR DESCRIPTION
**Pull request by chamini2**
This will avoid leaving the lock in an unexpected state

fixes #6

### CODE REVIEW

**Comment by chamini2** [comment]
From the [docs](https://docs.github.com/en/actions/managing-workflow-runs/canceling-a-workflow#steps-github-takes-to-cancel-a-workflow-run)

File: action.yml [hunk 1]
```diff
@@ -36,4 +36,6 @@ runs:
     ARG_DEBUG: ${{ inputs.debug }}
   entrypoint: '/scripts/lock.sh'
   post-entrypoint: '/scripts/unlock.sh'
+  if: ${{ always() }}
```
Comment (chamini2) [inline action.yml:39]: I only see documented `post-if` https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#post-entrypoint

Maybe we have to remove this line

Comment (chamini2) [reply action.yml:39]: And it seems it handles it well enough: https://github.com/ben-z/gh-action-mutex/blob/bc97a5192de56202e1b5d1d3eeca2ecbbef8523c/rootfs/scripts/utils.sh#L101-L105

File: action.yml [hunk 2]
```diff
@@ -36,4 +36,5 @@ runs:
     ARG_DEBUG: ${{ inputs.debug }}
   entrypoint: '/scripts/lock.sh'
   post-entrypoint: '/scripts/unlock.sh'
+  post-if: ${{ always() }}
```
Comment (ben-z) [inline action.yml:39]: I'm pretty sure this is already the default. The current behaviour is already "if a step fails, the lock gets cleaned up".

Comment (ben-z) [reply action.yml:39]: Here's the doc: https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runspost-if

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -53,4 +53,32 @@ jobs:
       - run: |
           echo "I am protected!"
           sleep 5
+          
+  always_unlock_test_1:
```
Comment (ben-z) [inline .github/workflows/main.yml:57]: These are good tests!

Would this job show up as failed? I'm not sure how to best test this while still having the job pass. Perhaps we can run the check in a subsequent step?

**Review by ben-z — CHANGES_REQUESTED** [review]
I think the change is already the default behaviour. The test is good, though we might want to check to see how we can have the pipeline pass while still test for this behaviour.

**Comment by ben-z** [comment]
The workflow didn't run, am I configuring the workflow properly? Perhaps it should be `pull_request_target` instead of `pull_request`? https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/

**Comment by ben-z** [comment]
Oh I see I don't run the PR on pull requests at all 😆 https://github.com/ben-z/gh-action-mutex/blob/271919c55332938179960557fcf28c8fa78be691/.github/workflows/main.yml#L1

File: action.yml (hunk 2 above)
Comment (ben-z) [reply action.yml]: I think the difference is in “canceling” vs “failing”. When we cancel a job I’m pretty sure the whole pipeline just gets terminated. When a step fails the pipeline keeps running, which is when the post jobs are triggered

**Comment by ben-z** [comment]
> Here you have an example of this failing: https://github.com/fal-ai/fal/runs/6264956528?check_suite_focus=true
> 
> Since we had many in the same workflow waiting for the lock, they came to unlock in an unexpected order and some of them just raised "we don't have the lock" errors but then when they did have it, they already had stopped.

I can't seem to find the "we don't have the lock" error. Can you point me to it?

**Comment by chamini2** [comment]
image

**Comment by chamini2** [comment]
Here you can see the use case in action: 3 waiting jobs are cancelled at the same time and we handle it fine: https://github.com/fal-ai/fal/runs/6601366282?check_suite_focus=true

**Comment by ben-z** [comment]
Oh I see! You are right, the post job assumes that we have the lock. If the post job runs before we even obtain the lock, it's going to fail.

Good changes. I will look again

File: rootfs/scripts/utils.sh [hunk 4]
```diff
@@ -75,40 +75,45 @@ wait_for_lock() {
...
 }
 
-# Wait for the lock to become available
+# Remove from the queue, when locked by it or just enqueued
 # args:
 #   $1: branch
 #   $2: queue_file
 #   $3: ticket_id
-unlock() {
+unqueue() {
```
Comment (ben-z) [inline rootfs/scripts/utils.sh:89]: Can we rename this to `dequeue`? I think that's the more common word for this action

File: rootfs/scripts/utils.sh [hunk 5]
```diff
@@ -75,40 +75,45 @@ wait_for_lock() {
...
 
-
-	echo "[$__ticket_id] Unlocking"
-
 	update_branch $__branch
 
-	if [ "$(cat $__mutex_queue_file | head -n 1)" != "$__ticket_id" ]; then
-		1>&2 echo "We don't have the lock! Ticket ID: $__ticket_id. Mutex file:"
-		cat $__mutex_queue_file
+	if [ "$(cat $__queue_file | awk NF | head -n 1)" != "$__ticket_id" ]; then
```
Comment (ben-z) [inline rootfs/scripts/utils.sh:98]: I'd reverse the logic here, change `!=` to `==` and swap the if and else blocks

**Comment by chamini2** [comment]
Changes done @ben-z!

**Review by ben-z — APPROVED** [review]
LGTM! Thanks for the patch!

**Comment by ben-z** [comment]
I had to remove the test because I don't see a clear way for them to have a passing status and still test what we want to test: https://github.com/ben-z/gh-action-mutex/commit/54229dd30dfc55e4dbb705e0c7dfd27e511fc6e6

**Comment by chamini2** [comment]
Understood, @ben-z. Thanks for merging. Will you do an alpha release with this?

**Comment by ben-z** [comment]
True, good idea! https://github.com/ben-z/gh-action-mutex/releases/tag/v1.0-alpha-5

**Pull request merged**